Add possibility to archive project

leanil opened the pull request:
Store archived status for each tag in a new table.
Archived projects are not visible on UI, unless searching for them or checkbox is ticked.
    
Resolves #31.

leanil commented on bot/src/slack.js:
```diff
@@ -45,6 +59,8 @@ const updateMessage = async (event) => {
   await upsertReport({ts, message})
   await clearTags(ts)
 
+  if (await handleArchiveCommand(event.message, event.ts)) return
```
The ts in `event.message` is when the message was originally sent.
I need to update the archived status with the ts of the modification.

leanil commented on client/src/components/leftPanel/LeftPanel.js:
```diff
@@ -22,15 +25,12 @@ class LeftPanel extends Component {
-          placeholder="search"
-        />
+        <QuickSearch value={searchString} onChange={this.onSearchChange} />
+        <ShowArchivedBox label="show archived projects" isChecked={showArchived} onChange={this.onShowArchivedChange} />
```
According to the issue, this box should be at the bottom.
But I didn't find a way to prevent the tags from going under the box,
and putting it on the top feels more consistent, as the controls are in one place.
![bottom](https://user-images.githubusercontent.com/5972402/74099728-f0433c00-4b26-11ea-8a6c-58fa86a660b1.jpg)
![image](https://user-images.githubusercontent.com/5972402/74099744-2680bb80-4b27-11ea-8fa8-7cb34ffe9bca.png)

paldys replied:
>  But I didn't find a way to prevent the tags from going under the box,

There is always a way. :slightly_smiling_face: 

What you are saying is not wrong (controls in one place), but I don't think it looks nice and from UX perspective what and how Tomas suggested could work better. Where your solution fails is the search. Search should work with archived projects too always, so it would not be consistent with what the user sees with the checkbox.

He wrote to have an icon/text on the end which would collapse/display archived messages. This could be hidden when you are searching. Also, I believe some visual extinction (an icon? ) for the archived projects would be also nice.

paldys commented on migrations/20200205161120_create_archived_table.js:
```diff
@@ -0,0 +1,24 @@
+      INSERT INTO archived(tag,is_archived)
+      SELECT DISTINCT tag, FALSE
+      FROM tag;
+  `)
```
Is there a reason you went with a completely new table instead of using `tag` table? There is a one-to-one relationship between them, isn't there?

leanil replied:
In the `tag` table, the tag is not unique, as each tag can be present on many reports.

paldys replied:
Oh.. I see.

paldys commented on client/src/components/leftPanel/LeftPanel.js:
```diff
@@ -6,14 +6,17 @@ import './LeftPanel.scss'
   }
 
-  handleChange = (event) => this.setState({ searchString: event.target.value })
+  onSearchChange = (event) => this.setState({ searchString: event.target.value })
```
Please, move refactors into a separate commit.

leanil committed:
Move quick search to separate component

leanil committed:
Add icon to mark archived projects

leanil committed:
Indicate processed command by a check mark

leanil committed:
Rename tag table to tagged

The new name describes the purpose of the table better:
to store the relation between reports and tags.

leanil committed:
Add possibility to archive project

Bot listens for (un)archive :__tag1: :__tagn: commands in messages.
Server stores archived status for each tag in a new table.
Archived projects are not visible on the UI, unless we search them
or the checkbox is ticked.

We will have multiple commands behaving similarly (see #32),
so abstact over these similarities:
 - the command strarts with a specific string
 - it performs some action based on the tags in the message
I've used regex to identify the command instead of string for
greater flexibilty.

Resolves #31.

leanil commented on server/src/index.js:
```diff
@@ -95,7 +95,7 @@ app.get('/api/reports-by-tags/:tag', authorize, async (req, res) => {
   const authors = reports.map((r) => r.user)
   const replyAuthors = [].concat(...reports.map((t) => t.replies.map((r) => r.user)))
-  const allMentions = 
+  const allMentions =
```
This will disappear when I rebase.

paldys commented on client/src/components/leftPanel/LeftPanel.js:
```diff
@@ -6,37 +6,58 @@ import './LeftPanel.scss'
+            placeholder="search"
+          />
+          {!searchString && <ShowArchivedBox label="show archived projects" isChecked={showArchived}
+                                             onChange={this.onShowArchivedChange} />}
```
(nitpick) Is eslint forcing you to do this formatting? If not I would prefer the following:
```
{!searchString && (
	<ShowArchivedBox
		label="show archived projects"
		isChecked={showArchived}
		onChange={this.onShowArchivedChange}
	/>
)}
```

paldys commented on client/src/components/leftPanel/LeftPanel.js:
```diff
@@ -6,37 +6,58 @@ import './LeftPanel.scss'
+          <ul>
+            {tags
+              .filter(({tag, isArchived}) => tag.toLowerCase().includes(lowerCaseSearchString) &&
+                                             (searchString || !isArchived || showArchived))
```
(nitpick) again if not eslint is forcing you to do this, let's go with:
```
  tags
    .filter(
      ({tag, isArchived}) =>
        tag.toLowerCase().includes(lowerCaseSearchString) &&
        (searchString || !isArchived || showArchived),
    )
    .map((tag) => <Tag key={tag.tag} tag={tag} />)
```

paldys commented on migrations/20200212110207_create_tag_table.js:
```diff
@@ -0,0 +1,28 @@
+        .comment('Emoji identifying some project, like :__bee:.')
+      table
+        .boolean('is_archived')
+        .comment('Indicates whether the project is hidden by default.')
```
Let's stick with:
```suggestion
        .comment('Indicates whether the project is archived.')
```

paldys reviewed (commented):
Looks good, just a few nitpicks.

Also, please fix _Add possibility to archive project_ commit's message, there are few grammar errors (e.g sentence _Store archived..._) and you address some things there which are fixed only in later commits (e.g. _give feedback on recognized commands_).

leanil commented:
Updated the icon and aligned it vertically.

leanil commented:
Display commands as regular reports.

paldys reviewed (approved):
:+1:
